modules/consensus/synchronize.go

```diff
@@ -175,39 +176,72 @@ func (cs *ConsensusSet) managedReceiveBlocks(conn modules.PeerConn) (returnErr e
 
 	// Read blocks off of the wire and add them to the consensus set until
 	// there are no more blocks available.
-	moreAvailable := true
-	for moreAvailable {
-		// Read a slice of blocks from the wire.
-		var newBlocks []types.Block
-		if err := encoding.ReadObject(conn, &newBlocks, uint64(MaxCatchUpBlocks)*types.BlockSizeLimit); err != nil {
-			return err
-		}
-		if err := encoding.ReadObject(conn, &moreAvailable, 1); err != nil {
-			return err
-		}
+	blocksChan := make(chan []types.Block, 1000)
+	var processingFailed int32
+	var downloadErr, acceptErr error
+	var wg sync.WaitGroup
 
-		// Integrate the blocks into the consensus set.
-		for _, block := range newBlocks {
-			stalled = false
-			// Call managedAcceptBlock instead of AcceptBlock so as not to broadcast
-			// every block.
-			acceptErr := cs.managedAcceptBlock(block)
-			// Set a flag to indicate that we should broadcast the last block received.
-			if acceptErr == nil {
-				chainExtended = true
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		moreAvailable := true
+		for moreAvailable && atomic.LoadInt32(&processingFailed) == 0 {
+			// Read a slice of blocks from the wire.
+			var newBlocks []types.Block
+			if err := encoding.ReadObject(conn, &newBlocks, uint64(MaxCatchUpBlocks)*types.BlockSizeLimit); err != nil {
+				downloadErr = err
+				close(blocksChan)
+				return
 			}
-			// ErrNonExtendingBlock must be ignored until headers-first block
-			// sharing is implemented, block already in database should also be
-			// ignored.
-			if acceptErr == modules.ErrNonExtendingBlock || acceptErr == modules.ErrBlockKnown {
-				acceptErr = nil
+			if err := encoding.ReadObject(conn, &moreAvailable, 1); err != nil {
+				downloadErr = err
+				close(blocksChan)
+				return
 			}
-			if acceptErr != nil {
-				return acceptErr
+			// Send blocks over the channel.
+			blocksChan <- newBlocks
+		}
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		func() {
+			for newBlocks := range blocksChan {
```

Comment:
Ah, I was testing with `go test` but Travis tests with `go test -v -race -tags='testing debug' -timeout=500s`. The later command succeeds here as well. So the "limited hashing power" issue is unrelated.